src/framework/execution/__tests__/dependencyGraph.test.ts

```diff
@@ -0,0 +1,692 @@
+    const args = executionHandlerSpy.mock.calls[0][0];
+    expect(args.jobState).toBeInstanceOf(FileSystemJobState);
+  });
+
+  test('should perform a flush of the jobState after a step was executed', async () => {
+    let jobStateFlushSpy;
+
+    const steps: IntegrationStep[] = [
+      {
+        id: 'a',
+        name: 'a',
+        types: [],
+        executionHandler: ({ jobState }) => {
+          jobStateFlushSpy = jest.spyOn(jobState, 'flush');
+        },
+      },
+    ];
+
+    const graph = buildStepDependencyGraph(steps);
+    await executeStepDependencyGraph(graph);
+
+    expect(jobStateFlushSpy).toHaveBeenCalledTimes(1);
+  });
+
+  test('should hold execution of dependent steps until current step has completed', async () => {
+    let resolveSpyA: Function;
+    const spyA = jest.fn(
+      (): Promise<void> =>
+        new Promise((resolve) => {
+          resolveSpyA = resolve;
+        }),
+    );
+    const spyB = jest.fn();
+
+    /**
+     * Graph:
+     * a - b
+     *
+     * In this situation, 'a' is the leaf node
+     * 'b' depends on 'a',
+     */
+    const steps: IntegrationStep[] = [
+      {
+        id: 'a',
+        name: 'a',
+        types: [],
+        executionHandler: spyA,
+      },
+      {
+        id: 'b',
+        name: 'b',
+        types: [],
+        dependsOn: ['a'],
+        executionHandler: spyB,
+      },
+    ];
+
+    const graph = buildStepDependencyGraph(steps);
+
+    const executionPromise = executeStepDependencyGraph(graph);
+
+    // wait for spyA to have been called
+    await waitForExpect(() => {
+      expect(spyA).toHaveBeenCalledTimes(1);
+    });
+
+    // ensure spyB has not been called yet
+    expect(spyB).toHaveBeenCalledTimes(0);
+
+    resolveSpyA();
+
+    // expect spyB to eventually be called
+    await waitForExpect(() => {
+      expect(spyB).toHaveBeenCalledTimes(1);
+    });
+
+    await executionPromise;
+  });
+
+  test('executes all leaf steps concurrently', async () => {
+    const resolvers: Function[] = [];
+
+    const testData = times(50, () => {
+      const spy = jest.fn(
+        (): Promise<void> =>
+          new Promise((resolve) => {
+            resolvers.push(resolve);
+          }),
+      );
+
+      const step: IntegrationStep = {
+        id: uuid(),
+        name: uuid(),
+        types: [],
+        executionHandler: spy,
+      };
+
+      return { step, spy };
+    });
+
+    const steps = testData.map((d) => d.step);
+    const spies = testData.map((d) => d.spy);
+
+    const graph = buildStepDependencyGraph(steps);
+
+    const executionPromise = executeStepDependencyGraph(graph);
+
+    // wait for all spies to have been called
+    await waitForExpect(() => {
+      spies.forEach((spy) => expect(spy).toHaveBeenCalledTimes(1));
+    });
+
+    resolvers.forEach((resolve) => resolve());
+
+    await executionPromise;
+  });
+
+  test('executes all leaf steps concurrently', async () => {
```

Comment:
This name is a duplicate of the test above, and even looks like it *might* be a complete copy of the previous test. 🤔

Reply:
owoh, idk how I did that